File: 2-String/387_firstUniqChar.py
import string
class Solution:
    def firstUniqChar1(self, s):
        """
        :type s: str
        :rtype: int
        """
        letters='abcdefghijklmnopqrstuvwxyz'
        index=[s.index(l) for l in letters if s.count(l) == 1]
        return min(index) if len(index) > 0 else -1

    def firstUniqChar2(self, s):
        return min([s.find(c) for c in string.ascii_lowercase if s.count(c)==1] or [-1])
    
sol = Solution()
s = 'leetcodel'
print(s.rpartition('t'))
ans = sol.firstUniqChar2(s)
print(ans)

File: 2-String/387_firstUniqChar.java
public class Solution{
    public int firstUniqChar(String s){
        int freq [ ] = new int[26];
        for (int i = 0; i < s.length(); i++)
            freq [s.charAt(i) - 'a'] ++;
        for (int i = 0; i < s.length(); i++)
            if(freq[s.charAt(i) - 'a'] == 1)
                return i;
        return -1;
    }

    public int firstUniqChar2(String s){
        if(s == null || s.length() == 0) return -1;
        int len = s.length();
        if(len == 1) return 0;
        char[] cc = s.toCharArray();
        int slow = 0, fast = 1;
        int[] count = new int[256];
        count[cc[slow]]++;
        while(fast < len){
            count[cc[fast]]++;
            while(slow < len && count[cc[slow]] > 1) slow++;
            if(slow >= len) return -1;
            if(count[cc[slow]] == 0){
                count[cc[slow]]++;
                fast = slow;
            }
            fast++;
        }
        return slow;
    }
}
